mat/cdense_arithmetic.go

```diff
@@ -0,0 +1,376 @@
+// Copyright ©2013 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	//	"fmt"
+	"math"
+
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/cblas128"
+	"gonum.org/v1/gonum/lapack/clapack128"
+)
+
+// Add adds a and b element-wise, placing the result in the receiver. Add
+// will panic if the two matrices do not have the same shape.
+func (m *CDense) Add(a, b CMatrix) {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br || ac != bc {
+		panic(ErrShape)
+	}
+
+	aU, _, _ := untransposeCmplx(a)
+	bU, _, _ := untransposeCmplx(b)
+	m.reuseAsNonZeroed(ar, ac)
+
+	if arm, ok := a.(*CDense); ok {
+		if brm, ok := b.(*CDense); ok {
+			amat, bmat := arm.mat, brm.mat
+			if m != aU {
+				m.checkOverlap(amat)
+			}
+			if m != bU {
+				m.checkOverlap(bmat)
+			}
+			for ja, jb, jm := 0, 0, 0; ja < ar*amat.Stride; ja, jb, jm = ja+amat.Stride, jb+bmat.Stride, jm+m.mat.Stride {
+				for i, v := range amat.Data[ja : ja+ac] {
+					m.mat.Data[i+jm] = v + bmat.Data[i+jb]
+				}
+			}
+			return
+		}
+	}
+}
```

Comment:
This would have something like the following as the non-fast path:

```
	m.checkOverlapMatrix(aU)
	m.checkOverlapMatrix(bU)
	var restore func()
	if aTrans != aConj && m == aU {
		m, restore = m.isolatedWorkspace(aU)
		defer restore()
	} else if bTrans != bConj && m == bU {
		m, restore = m.isolatedWorkspace(bU)
		defer restore()
	}

	for r := 0; r < ar; r++ {
		for c := 0; c < ac; c++ {
			m.set(r, c, a.At(r, c)+b.At(r, c))
		}
	}
```

(with `aTrans`, `aConj`, `bTrans` and `bConj` coming from the relevant `untransposeCmplx` call).

The other element-wise operations will have the analogous addition.

What we're doing here is making sure that if the destination overlaps with the src and there is a transposition that we make a temporary work place to put the result in when then gets copied back into the destination. Otherwise the partial work will clobber the source.

Reply:
@kortschak I'm not sure I understand this pattern. It looks like the fast path includes type assertions (i.e. `arm, ok := a.(*CDense); ok { /* CDense methods are now available */ }`. In what case would you fall through into the slow path? The input is, by definition, always going to be a `CMatrix` which, if I understand correctly, will always satisfy the assertion `a.(*CDense)`?

Reply:
It's true that `*CDense` is only only `CMatrix` implementation that we *currently* provide, but there will be others. There may be other implementations in the wild (unlikely, but possible). The design of the package is that we accept the interface, not the implementation.